File: src/hollywood/adapters/wikidata.py
from __future__ import annotations

import json
from pathlib import Path

import httpx

from ..config import HollywoodSettings
from ..models import (
    EntityAliasRow,
    EntityKind,
    EntityRow,
    IngestOptions,
    NormalizedBundle,
    RawPayload,
    SourceDefinition,
    json_dumps,
    make_stable_id,
)
from ..storage import HollywoodStorage
from .base import BaseAdapter

WIKIDATA_QUERY = """
SELECT ?item ?itemLabel ?occupationLabel ?imdb WHERE {
  VALUES ?occupation { wd:Q33999 wd:Q2526255 wd:Q28389 }
  ?item wdt:P31 wd:Q5;
        wdt:P106 ?occupation.
  OPTIONAL { ?item wdt:P345 ?imdb. }
  SERVICE wikibase:label { bd:serviceParam wikibase:language "en". }
}
LIMIT {limit}
"""


class WikidataAdapter(BaseAdapter):
    def __init__(self, source: SourceDefinition):
        super().__init__(source)

    def fetch_raw_payloads(
        self, settings: HollywoodSettings, options: IngestOptions
    ) -> list[RawPayload]:
        limit = options.limit or 25
        query = WIKIDATA_QUERY.format(limit=limit)
        response = httpx.get(
            self.source.default_urls[0],
            params={"query": query, "format": "json"},
            headers={
                "Accept": "application/sparql-results+json",
                "User-Agent": settings.user_agent,
            },
            timeout=settings.request_timeout_seconds,
            follow_redirects=True,
        )
        response.raise_for_status()
        return [
            RawPayload(
                payload_type="api_json",
                logical_id=f"wikidata-{limit}",
                body=response.content,
                content_type=response.headers.get("content-type", "application/json"),
                source_url=str(response.url),
                metadata={"query": query, "limit": limit},
                extension=".json",
            )
        ]

    def normalize_raw_records(
        self,
        settings: HollywoodSettings,
        storage: HollywoodStorage,
        run_id: str,
        raw_records: list[dict[str, object]],
    ) -> NormalizedBundle:
        bundle = NormalizedBundle()
        for record in raw_records:
            if str(record["payload_type"]) != "api_json":
                continue
            document = json.loads(Path(str(record["content_path"])).read_text(encoding="utf-8"))
            bindings = document.get("results", {}).get("bindings", [])
            for item in bindings:
                item_url = item.get("item", {}).get("value")
                label = item.get("itemLabel", {}).get("value")
                imdb_id = item.get("imdb", {}).get("value")
                occupation = item.get("occupationLabel", {}).get("value")
                if not item_url or not label:
                    continue
                qid = item_url.rsplit("/", 1)[-1]
                entity_id = make_stable_id("wikidata", qid)
                bundle.entities.append(
                    EntityRow(
                        entity_id=entity_id,
                        source_id=self.source.source_id,
                        external_id=qid,
                        entity_type=EntityKind.PERSON.value,
                        name=label,
                        canonical_name=label.casefold(),
                        license_class=self.source.license_class.value,
                        metadata_json=json_dumps({"occupation": occupation, "imdb_id": imdb_id}),
                    )
                )
                bundle.entity_aliases.append(
                    EntityAliasRow(
                        entity_alias_id=make_stable_id(entity_id, label),
                        entity_id=entity_id,
                        source_id=self.source.source_id,
                        alias=label,
                    )
                )
        return bundle

File: src/ingestion/adapters/wikidata.ts
import { readFileSync } from 'node:fs';

import type { RawRecordRow } from '../../domain/repositories/RawRecordRepository.js';
import { env } from '../../env.js';
import { emptyBundle, makeStableId } from '../models.js';
import type {
  EntityAliasRow,
  EntityRow,
  IngestOptions,
  NormalizedBundle,
  RawPayload,
  SourceDefinition,
} from '../models.js';
import type { Adapter } from './base.js';

const WIKIDATA_QUERY = `
SELECT ?item ?itemLabel ?occupationLabel WHERE {
  VALUES ?occupation { wd:Q33999 wd:Q2526255 wd:Q28389 }
  ?item wdt:P31 wd:Q5;
        wdt:P106 ?occupation.
  SERVICE wikibase:label { bd:serviceParam wikibase:language "en". }
}
LIMIT {limit}
`;

interface SparqlBinding {
  item?: { value?: string };
  itemLabel?: { value?: string };
  occupationLabel?: { value?: string };
}

export class WikidataAdapter implements Adapter {
  constructor(public source: SourceDefinition) {}

  async fetchRawPayloads(options: IngestOptions): Promise<RawPayload[]> {
    const limit = options.limit ?? 25;
    const query = WIKIDATA_QUERY.replace('{limit}', String(limit));
    const url = new URL(this.source.defaultUrls[0]!);
    url.searchParams.set('query', query);
    url.searchParams.set('format', 'json');

    const resp = await fetch(url, {
      headers: {
        Accept: 'application/sparql-results+json',
        'User-Agent': env.HOLLYWOOD_USER_AGENT,
      },
      signal: AbortSignal.timeout(env.HOLLYWOOD_REQUEST_TIMEOUT_SECONDS * 1000),
    });
    if (!resp.ok) throw new Error(`Wikidata request failed: ${resp.status}`);
    const body = Buffer.from(await resp.arrayBuffer());

    return [
      {
        payloadType: 'api_json',
        logicalId: `wikidata-${limit}`,
        body,
        contentType: resp.headers.get('content-type') ?? 'application/json',
        sourceUrl: resp.url,
        fetchedAt: new Date(),
        metadata: { query, limit },
        extension: '.json',
      },
    ];
  }

  async normalizeRawRecords(_runId: string, rawRecords: RawRecordRow[]): Promise<NormalizedBundle> {
    const bundle = emptyBundle();
    for (const record of rawRecords) {
      if (record.payloadType !== 'api_json') continue;
      const document = JSON.parse(readFileSync(record.contentPath, 'utf-8')) as {
        results?: { bindings?: SparqlBinding[] };
      };
      const bindings = document.results?.bindings ?? [];

      for (const item of bindings) {
        const itemUrl = item.item?.value;
        const label = item.itemLabel?.value;
        const occupation = item.occupationLabel?.value ?? null;
        if (!itemUrl || !label) continue;

        const qid = itemUrl.split('/').pop()!;
        const entityId = makeStableId('wikidata', qid);
        const entityRow: EntityRow = {
          entityId,
          sourceId: this.source.sourceId,
          externalId: qid,
          entityType: 'person',
          name: label,
          canonicalName: label.toLowerCase(),
          metadataJson: JSON.stringify({ occupation }),
        };
        bundle.entities.push(entityRow);

        const aliasRow: EntityAliasRow = {
          entityAliasId: makeStableId(entityId, label),
          entityId,
          sourceId: this.source.sourceId,
          alias: label,
        };
        bundle.entityAliases.push(aliasRow);
      }
    }
    return bundle;
  }
}
